stream/buffer.go

```diff
@@ -0,0 +1,59 @@
+package stream
+
+import (
+	"io"
+	"sync"
+)
+
+// Buffer is an in-memory stream which supports one writer and multiple readers.
+// The stream is backed by a byte slice and concurrency is handled by an RWLock.
+type Buffer struct {
+	buf []byte
+	rw  *sync.RWMutex
+}
+
+// bufferReader represents a single reader of a Buffer. It essentialy is an
+// index into the Buffer which indicates the current position from read data
+// from. Multiple instances of bufferReader can read concurrently from the
+// underlying Buffer
+type bufferReader struct {
+	buffer *Buffer
+	i      int64
+}
+
+// NewBuffer returns an empty Buffer backed by a byte slice of initial length 0
+// and default capacity
+func NewBuffer() Source {
+	return &Buffer{
+		buf: make([]byte, 0),
+		rw:  new(sync.RWMutex),
+	}
+}
+
+// Open returns a new reader into Buffer. Open may be called multiple times to
+// retrieve multiple readers. All readers may read concurrently
+func (s *Buffer) Open() (io.Reader, error) {
+	return &bufferReader{
+		buffer: s,
+		i:      0,
+	}, nil
+}
+
+func (s *bufferReader) Read(b []byte) (n int, err error) {
```

Comment:
I got a bit confused by the variable names while reading through the function. Personally I like having one character long receiver names but would have named the parameter `dst` and the receiver then `b` (because it's the *buffer*reader or `r` because it's the *reader*? Something like that). Also `n` is *okay* but maybe instead of `n` something like `bytesWritten` (such name is already used in the `main.go`). It's a personal thing, I like names that describe what they are used for (as far as it's still convenient to write that name).
```suggestion
func (r *bufferReader) Read(dst []byte) (int, error) {
```

Reply:
That's probably just me being lazy. Inspired from here: https://golang.org/src/bytes/reader.go?s=3877:3909#L40

We can change that for sure. Will probably batch them together in a single commit